keps/sig-network/2104-kube-proxy-lib/README.md

```diff
@@ -0,0 +1,280 @@
+tools.  This KEP distills the goals of such an interface and propose it's lifecycle and official support policy for sig-network.
+
+This distillation of a "KPNG like interface" was originally presented by tim hockins in a sig-network breakout session, informally in https://docs.google.com/presentation/d/1Y-tZ4fFC9L2NvtBeiIXD1MiJ0ieg_zg4Q6SlFmIax8w/edit?hl=en&resourcekey=0-SFhIGTpnJT5fo6ZSzQC57g#slide=id.g16976fedf03_0_221.
+
+## Motivation
+
+There have been several presentations, issues, and projects dedicated to reusing kube proxy logic while extending it to embrace
+different backend technologies (i.e. NFT, eBPF, openvswitch, and so on).  This KEP attempts to make a library which will facilitate
+this type of work.  
+
+A general solution to this problem is explored in the KPNG project (https://github.com/kubernetes-sigs/kpng/), which exhibits many properties
+that allow for such goals to be accomplished.  These are enabled by:
+
+- A Generic "*Diff store"* which provides a client side, in memory data model for performant, declarative, generic calculation of differences between the Kubernetes networking state space from one time point to another (i.e. a replacement for the implicit change tracking functionality in k/k/pkg/proxy https://github.com/kubernetes/kubernetes/blob/master/pkg/proxy/endpoints.go#L161).
+- A *library that simplifies consumption* of the Kubernetes network and its topology easy to consume for an individual node, abstracting away API Semantics (like topology) from underlying network routing rules.
+- *Definition of types* which have a minimal amount of boiler plate, can be created without using the Kubernetes API data model (and thus used to extend proxying behaviour to things outside of the Kubernetes pod, service, and endpoint semantics), and which can describe routing of Kubernetes VIPs (services) to endpoints in a generic way that is understandable to people who don't work on Kubernetes on a day-to-day basis.
+
+### Goals
+
+- Build a vendorable repository "kube-proxy-lib" which can be used to make new kube-proxy's.
+- Exemplify the use of such a repository in a mock "backend" which uses the repository to process and respond to changes in the Kubernetes networking state.
+- Define a policy around versioning and releasing of "kube-proxy-lib".
+- Create a CI system that runs in test-grid which tests kube-proxy-lib compatibility with the latest Kubernetes releases continuously, and which can be used to inform API changes for regressions that break "kube-proxy-lib".
+- Enable the eventual *replacement* of the k/k/pkg/proxy serviceChangeTracker and endpointsChangeTracker related caching structures inside of in-tree kube proxy with this generic library.
+
+### Non-Goals
+
+- Rewrite or decouple the core , in-tree linux Kubernetes kube-proxy implementations, which are relied on by too many users to be tolerant to major changes.
```

Comment:
This is incompatible with the goal of replacing ServiceChangeTracker/EndpointsChangeTracker above.

In fact, I think that rebasing the iptables, ipvs, and winkernel implementations on top of this library is a _requirement_, not a non-goal. We don't want to be maintaining two separate implementations of "proxy business logic". (The entire point of this project is to reduce the maintenance burden of people keeping up with "proxy business logic".)

Certainly we need to be careful to not destabilize the existing implementations, but it seems to me that the way to do that is to use the existing shared proxy code (ServiceChangeTracker, etc) as the initial seed of the library, and replace it with new code bit by bit (ensuring that unit, integration, and e2e tests pass all along the way).

Reply:
> In fact, I think that rebasing the iptables, ipvs, and winkernel implementations on top of this library is a requirement, not a non-goal.

I agree here, however some initial changes like adding shared bits for starting and stopping API server handlers and watchers will also need to be ported i.e something like the [Provided interface](https://github.com/kubernetes/kubernetes/blob/master/pkg/proxy/types.go#L29) right?

 IMO the very minimum the first POC/ Seed should still provide external users the ability to start writing new backends quickly not just replace the client side caching mechanisms.